lxd/db/instances.go

```diff
@@ -299,59 +299,62 @@ var ErrInstanceListStop = fmt.Errorf("search stopped")
 
 // InstanceList loads all instances across all projects and for each instance runs the instanceFunc passing in the
 // instance and it's project and profiles. Accepts optional filter argument to specify a subset of instances.
-func (c *Cluster) InstanceList(filter *InstanceFilter, instanceFunc func(inst Instance, project api.Project, profiles []api.Profile) error) error {
-	var instances []Instance
-	projectMap := map[string]api.Project{}
-	projectHasProfiles := map[string]bool{}
-	profilesByProjectAndName := map[string]map[string]api.Profile{}
+func (c *Cluster) InstanceList(filter *cluster.InstanceFilter, instanceFunc func(inst InstanceArgs, project api.Project, profiles []api.Profile) error) error {
+	var instances []InstanceArgs
+	projectsByName := map[string]api.Project{}
 
 	if filter == nil {
-		filter = &InstanceFilter{}
+		filter = &cluster.InstanceFilter{}
 	}
 
 	// Retrieve required info from the database in single transaction for performance.
 	err := c.Transaction(context.TODO(), func(ctx context.Context, tx *ClusterTx) error {
-		var err error
-		instances, err = tx.GetInstances(*filter)
+		dbInstances, err := cluster.GetInstances(ctx, tx.tx, *filter)
 		if err != nil {
 			return fmt.Errorf("Failed loading instances: %w", err)
 		}
 
-		projects, err := cluster.GetProjects(ctx, tx.tx, cluster.ProjectFilter{})
-		if err != nil {
-			return fmt.Errorf("Failed loading projects: %w", err)
-		}
-
-		// Index of all projects by name and record which projects have the profiles feature.
-		for _, project := range projects {
-			apiProject, err := project.ToAPI(ctx, tx.tx)
+		instances = make([]InstanceArgs, 0, len(dbInstances))
+		for _, instance := range dbInstances {
+			// Get expanded instance information.
+			instanceArgs, err := InstanceToArgs(ctx, tx.tx, &instance)
 			if err != nil {
 				return err
 			}
 
-			projectMap[project.Name] = *apiProject
-			projectHasProfiles[project.Name] = shared.IsTrue(apiProject.Config["features.profiles"])
+			// Add an empty api.Project so we know this is an instance project.
+			_, ok := projectsByName[instance.Project]
+			if !ok {
+				projectsByName[instance.Project] = api.Project{}
+			}
+
+			instances = append(instances, *instanceArgs)
 		}
 
-		profiles, err := cluster.GetProfiles(ctx, tx.tx, cluster.ProfileFilter{})
+		projects, err := cluster.GetProjects(ctx, tx.tx, cluster.ProjectFilter{})
```

Comment:
Can we avoid this `GetProjects` call by changing the for loop below to iterate over the project names generated in `projectsByName` above? Seeing as we are just populating those entries here.

Reply:
Doing this would result in more database hits as we would be talking to SQL once for each iteration of `projectsByName` rather than just taking everything at once and filtering it in `go`.

Reply:
Right I see, so you're saving queries by potentially transferring more data from the DB (probably the right choice).
Makes sense.